### ISSUE TITLE
double free or corruption

### ISSUE THREAD

**Issue by dschwoerer** [boutproject/BOUT-dev#342]
If I run the testsuite I get plenty of output from exits.

This is new, and quite annoying.

Before I had this issue only with unclean exits, but it seems now this is also with (supposedly) clean exits.

```
Error norm g: l-2 0.000094 l-inf 0.000147
Running with mesh:ny=256 mesh:dy=0.0245436926062 nout=1 timestep=1
*** Error in `./wave': corrupted double-linked list: 0x00000000015b07d0 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x77de5)[0x7f2fa12d0de5]
/lib64/libc.so.6(+0x80936)[0x7f2fa12d9936]
/lib64/libc.so.6(cfree+0x4c)[0x7f2fa12dc78c]
./wave[0x40da20]
/lib64/libc.so.6(__libc_start_main+0xf0)[0x7f2fa1279580]
./wave[0x40d809]
======= Memory map: ========
```

[.... plenty of output .....]

```
*** Error in `./2fluid': double free or corruption (!prev): 0x0000000002818b60 ***
*** Error in `./2fluid': double free or corruption (!prev): 0x00000000026b2af0 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x77de5)[0x7ff8d30e3de5]
======= Backtrace: =========
/lib64/libc.so.6(+0x8022a)[0x7ff8d30ec22a]
/lib64/libc.so.6(+0x77de5)[0x7f8a768dbde5]
/lib64/libc.so.6(cfree+0x4c)[0x7ff8d30ef78c]
./2fluid[0x4c66ed]
./2fluid[0x411049]
/lib64/libc.so.6(__libc_start_main+0xf0)[0x7ff8d308c580]
./2fluid[0x4c66ed]
./2fluid[0x411049]
/lib64/libc.so.6(__libc_start_main+0xf0)[0x7f8a76884580]
./2fluid[0x40d899]
======= Memory map: ========
```

**Comment by d7919** [comment]
Is this in next or master?

**Comment by dschwoerer** [comment]
Hi, this is in next.

No, couldn't find any updates of gcc or glibc in the logs, gcc is version 5.3.1

**Comment by ZedThree** [comment]
What MPI library are you using?

**Comment by dschwoerer** [comment]
MPICH, but I am unable to reproduce this issue ... not sure what happened ....

**Comment by ZedThree** [comment]
I get segfaults in strange places with MPICH and gnu 6 that go away if I
switch to openmpi. Very difficult to diagnose. It would be good to get to
the bottom of this!

**Comment by dschwoerer** [comment]
Appeared again (v4 RC with minor changes)
```
(gdb) bt
#0  0x00007ffff25c8765 in raise () from /lib64/libc.so.6
#1  0x00007ffff25ca36a in abort () from /lib64/libc.so.6
#2  0x00007ffff26113c0 in free_check () from /lib64/libc.so.6
#3  0x00007ffff26154b7 in free () from /lib64/libc.so.6
#4  0x000000000047647a in __gnu_cxx::new_allocator<Array<double>::ArrayData*>::deallocate (__p=0x7c41c0, this=0x7c4af8)
    at /usr/include/c++/6.2.1/ext/new_allocator.h:110
#5  std::allocator_traits<std::allocator<Array<double>::ArrayData*> >::deallocate (__n=<optimized out>, __p=0x7c41c0, __a=...)
    at /usr/include/c++/6.2.1/bits/alloc_traits.h:442
#6  std::_Vector_base<Array<double>::ArrayData*, std::allocator<Array<double>::ArrayData*> >::_M_deallocate (__n=<optimized out>, __p=0x7c41c0, this=0x7c4af8)
    at /usr/include/c++/6.2.1/bits/stl_vector.h:178
#7  std::vector<Array<double>::ArrayData*, std::allocator<Array<double>::ArrayData*> >::_M_emplace_back_aux<Array<double>::ArrayData* const&> (this=0x7c4af8, 
    __args#0=@0x7fffffffce88: 0x7c4b20) at /usr/include/c++/6.2.1/bits/vector.tcc:438
#8  0x00000000004767f6 in std::vector<Array<double>::ArrayData*, std::allocator<Array<double>::ArrayData*> >::push_back (__x=@0x7fffffffce88: 0x7c4b20, 
    this=<optimized out>) at /usr/include/c++/6.2.1/bits/stl_vector.h:924
#9  Array<double>::release (this=<optimized out>, d=<optimized out>) at ../../include/bout/array.hxx:303
#10 0x000000000047973c in Array<double>::~Array (this=0x790180 <n+160>, __in_chrg=<optimized out>) at ../../include/bout/array.hxx:77
#11 Field3D::~Field3D (this=0x7900e0 <n>, __in_chrg=<optimized out>) at field3d.cxx:128
#12 0x00007ffff25cd258 in __run_exit_handlers () from /lib64/libc.so.6
#13 0x00007ffff25cd2a5 in exit () from /lib64/libc.so.6
#14 0x00007ffff3084689 in MPL_exit () from /usr/lib64/mpich/lib/libmpi.so.12
#15 0x00007ffff303c6bc in MPID_Abort () from /usr/lib64/mpich/lib/libmpi.so.12
#16 0x00007ffff2fa5e91 in PMPI_Abort () from /usr/lib64/mpich/lib/libmpi.so.12
#17 0x000000000046f6a7 in main (argc=<optimized out>, argv=<optimized out>) at ../../include/boutmain.hxx:138
```
I throw an exception in physics_init - the cleanup afterwards causes the double free / corruption.

The field destroyed is a global Field3D, which has been initialized.

The field looks like this:
```
$2 = {<Field> = {_vptr.Field = 0x78a6f8 <vtable for Field3D+16>, bndry_xin = true, bndry_xout = true, bndry_yup = true, bndry_ydown = true}, <FieldData> = {
    _vptr.FieldData = 0x78a7e0 <vtable for Field3D+248>, bndry_op = std::vector of length 0, capacity 0, boundaryIsCopy = false, boundaryIsSet = true, 
    bndry_op_par = std::vector of length 0, capacity 0, bndry_generator = std::map with 0 elements}, background = 0x0, fieldmesh = 0x7b5750, nx = 5, ny = 8, nz = 5, 
  data = {ptr = 0x7c4b20, static store = {_M_t = {
        _M_impl = {<std::allocator<std::_Rb_tree_node<std::pair<int const, std::vector<Array<double>::ArrayData*, std::allocator<Array<double>::ArrayData*> > > > >> = {<__gnu_cxx::new_allocator<std::_Rb_tree_node<std::pair<int const, std::vector<Array<double>::ArrayData*, std::allocator<Array<double>::ArrayData*> > > > >> = {<No data fields>}, <No data fields>}, _M_key_compare = {<std::binary_function<int, int, bool>> = {<No data fields>}, <No data fields>}, _M_header = {_M_color = std::_S_red, 
            _M_parent = 0x7b9940, _M_left = 0x7b9940, _M_right = 0x7c4ad0}, _M_node_count = 2}}}, static use_store = true}, location = CELL_CENTRE, deriv = 0x7c3920, 
  yup_field = 0x0, ydown_field = 0x0}
```

**Comment by ZedThree** [comment]
What MPICH version are you using? Can you try with a different MPI implementation, or a different compiler?

**Comment by ZedThree** [comment]
As far as I can tell the problem is in `BoutComm::getComm()` and only appears with GCC >= 5.0 and MPICH 3 (?). I can't reproduce with GCC < 5.0 or other MPIs (including MPICH 2).

You're seeing it in `MPI_Abort(BoutComm::get(), 1);`, I'm seeing segfaults in `MPI_Barrier(BoutComm::get());` in `BoutFinalise`. I've previously seen crashes in MPI functions that originate in `BoutComm::get()`. I'm currently suspecting `MPI_Comm_dup` returning garbage, but I can't prove it yet.

Which version of MPICH are you using? Do you have a minimal example I can look at?

**Comment by ZedThree** [comment]
Oh, there was a bug in MPICH 3.1 (which I'm using): https://trac.mpich.org/projects/mpich/ticket/2217

I'll try out MPICH 3.2 and see if that fixes it. @dschwoerer Please could you do the same, if you're currently using 3.1?

**Comment by dschwoerer** [comment]
This is really driving me crazy :confused: 
I tried with both mpich 3.1 and 3.2 - I cannot reproduce this issue in ether.
I really should investigate it in more detail immediately when I have them.
I am wondering right know, whether it may be some glitch with recompiling only part of the library, I think this was always the case when I had the issues, and when I recompiled everything later to reproduce, it was always gone ...
May as well be something different ...

In my case it was in the deconstruction of Field3D, so I don't suspect a direct involvement of MPI.
